### ISSUE TITLE
Folder grouping proposal

### ISSUE THREAD

**Issue by Getron** [syncthing/syncthing#2070]
I have a scenario, where i sync folders from my customers to, not just my private folders, on my server.

Now, i have a really mess on folders, all the folders are together, on one list on the left side, my private, and from all my customers.

The idea is to be able to add groups to (maybe, subgroups on groups to, if possible), on the left folder side, and easy to able to sort the sync folders, like three, or somthing in this direction.

Then, if i have, 50 customers, and sync data of all 50 customers, i be able to sort them up in groups. No more messy (just one list) with folders then.

Should be somthing like this (on the left folder side):

- Swiss Machines
  - My PC nr.1
    - Documents
    - Music
    - Photos
  - My PC nr.2
    - Docus
    - Videos
  - Customer 1
    - Documents
    - Photos
  - Customer 2
    - Documents
    - Work files
- Machines on other lands
  - Work PC nr1
    - Documents
    - Pics
 - Server 1
   - Backup1
   - Backup2

... and so on... hope, you understand the idea.

I think, this would help out to sort, and easy to find the folders in a lists.

What you think? Is this doable, and worth to have? I think, many people out there would apreciate to have somthing like this, to able to easy, and excelently sort the sync folder from many machines, server, or whatever.

Thanks in advance to devs, if this enhancement would be added to.

**Comment by uok** [comment]
Sounds like a helpful enhancement. 

My suggestion is to keep it to one level (to keep the GUI elements that can collapse/expand minimal and also because bootstrap framework adds some padding here and there), so in your case you would maybe have 3 groups (e.g. separating Swiss, yours, other countries). For further sorting you could use a simple folder prefix like "MyPC1_documents", etc. The "group label" itself could be done maybe with a text field in each folder edit dialog. 

When Syncthing reads the folder config it simply adds a collapsible element if a group label exists and then puts the folders inside. if all group label fields of all folders are empty everything looks like it does now

**Comment by wweich** [comment]
You need to prefix (or suffix) the folder ids anyway because they have to be unique (although e.g #996 could change that).

So naming the folders e.g. Swiss_MyPC1_Documents and OtherLands_WorkPC1_Documents and so on will group them inside the plain list.

**Comment by Getron** [comment]
@uok  Yes, GUI should be the same, expand and colapse the groups, subgroups or folders. It should be somthing like the Wordpress do with menu on admin side. If you known how wordpress works, you understand what i mean. Helpful functionality.

@wweich Yes, all the system of syncthing should be the same, just grouping them, to be organized, and to able to find, or sort them properly.

**Comment by uok** [comment]
I don't think it should be automatic as some people prefer a simple list. Although I have to admin that I'm not sure how many people will use this feature because according to the [stats](https://data.syncthing.net) the average user has ~10 folders

still, here is a quick mockup :smile: 
(folder **#**1 is a regular, non-grouped folder above a set of grouped folders)

collapsed
![group-collapsed](https://cloud.githubusercontent.com/assets/9367272/8767737/6c2d0874-2e66-11e5-87f6-d67734bdad4c.png)
and all expanded
![group-expanded](https://cloud.githubusercontent.com/assets/9367272/8767740/752e7b6a-2e66-11e5-9afa-29d8fad9f256.png)
if you like to fiddle with the fiddle: https://jsfiddle.net/h31cm4p3/1/
HTML-wise this is ready to copy/paste into Syncthing, only some GO/angular-magic is required while reading the folder config to add the html tags at the correct position

**Comment by Getron** [comment]
@uok Yes, somthing like this would be great to.

Unfortunatly, i can't do it myself. If i can, i would, but, i can't.

I apreciate, if someone would add this, a great thing. And +- 10 folder for average user, with the new technology, you can just thinking, in one family, how many mobiles phone exist? and if any of these would syncing, the +-10 folder are quickly overnumbered, hehe. We have not count the synced Computer folders here, hehe... Just my family have, every person 5-10 folders to share. My brother, 5-10, father pc, mother pc, my sons pc, my wife pc, and so on... Trust me, many people would be apreciate to have this sorting to. If not now, in a year, or two...

**Comment by Getron** [comment]
Has anyone idea how to implement this?

**Comment by lkwg82** [comment]
Later on when folder can have labels ... there could be an ui rule to interpret dots (or other chars) in labels as separators to introduce a visual hierarchy:

- folderID A (Label: swiss#photos)
- folderID B (Label: swiss#baby)
- folderID C (Label: germany#bavaria)
- folderID D (Label: germany#berlin)

swiss
- photos
- baby

germany
- bavaria
- berlin

What about that?

**Comment by Nejaa** [comment]
I'm also 100% needing this feature

**Comment by lkwg82** [comment]
@Nejaa Take some time and propose a PR or a algorithm.

**Comment by vincentardern** [comment]
This should be fairly simple to do dynamically with JQuery I think. Any obvious problems? I might have a go if nobody can think of a major roadblock.

**Comment by AudriusButkevicius** [comment]
Probably angular that than jquery

**Comment by whoabuddy** [comment]
Nothing like reviving an old issue, but this would be an amazing feature! I chose to manage less folders overall since I am the only one accessing the data, but over time I saw the need (and desire) to set up a LOT more folders for various reasons.

My big two reasons were more granular control of who the data is shared with, which allows me to share folders with partners and clients, and segmenting large files or directories, which allows me to choose which folders I _really_ need on devices that have less storage or bandwidth.

My current setup is below, and I am going to dive in further to see if there is anything I can contribute. I am up to 17 folders so far across 4 devices, and I've run into some interesting challenges along the way.

I've also tried using naming conventions but I agree that grouping feels like a much more natural solution to the problem. Less clutter is better, and it makes it faster to get to a specific folder within a group.

If I set up all of the folders _I would like to use_, then my total would be between 40-50 folders. This did not seem manageable when I first started, but I've learned a lot since then, too. :)

----------

### Current Setup

__Business Folders__

- Business _(could be further split per business)_
- Consulting _(could be further split per client)_
- Development _(could be further split per project)_
- Website _(could be further split per site)_

__Personal Folders__

- Listen
- Notes
- Passwords _(encrypted kdbx file, don't get too excited!)_
- Personal

__Device Folders__

- Phone-XX-DCIM
- Phone-XX-Downloads
- Phone-XX-Pictures
- Phone-YY-DCIM
- Phone-YY-Downloads
- Phone-YY-Pictures
- Phone-ZZ-DCIM
- Phone-ZZ-Downloads
- Phone-ZZ-Pictures

Where XX, YY, and ZZ represent the phone owner's initials.

**Comment by whoabuddy** [comment]
_Disclaimer: the more confused I get, the more I go into detail, so prepare for a long comment! ☕_

I love this idea, but I am a bit at a loss on how to implement it. I gave it my best shot in the outline below, and I would love some additional input from the more familiar: @uok @lkwg82 @AudriusButkevicius.

If we can hash out a basic plan I am happy to try and make the updates.

-----

Looking at `syncthing/gui/default/index.html`, I can see the folder list and the Angular function that repeats each individual folder.

```
<!-- Folder list (top left) -->
<div class="col-md-6" aria-labelledby="folder_list" role="region" >
  <h3 id="folder_list" translate>Folders</h3>
  <div class="panel-group" id="folders">
    <div class="panel panel-default" ng-repeat="folder in folderList()">
```

Which led me to `syncthing/gui/default/syncthing/app.js` and the `folderList()` function.

```
function folderList(m) {
    var l = [];
    for (var id in m) {
        l.push(m[id]);
    }
    l.sort(folderCompare);
    return l;
}
```

Cool, now I know how the folders are populated in the template, and the first idea I looked at was using a special label name per @lkwg82.

This definitely sounds like the easiest route without affecting too much code, but I'm still trying to wrap my head around the best way to implement it.

Right now the `ng-repeat="folder in folderList()"` function will:

- create a button for the heading
  - create a progress bar based on sync percentage (top color bar)
  - create the h4 for the panel title
    - create the folder icon
    - create the folder status
    - create the title text
- create the collapsed div with folder info
  - create/populate the folder info

-----

Let's use the example of the phone folders I provided in my last comment, but change the delimiter per lkwg82's suggestion:

__Folder Labels__

```
Phone#XX#DCIM
Phone#XX#Downloads
Phone#XX#Pictures
Phone#YY#DCIM
Phone#YY#Downloads
Phone#YY#Pictures
Phone#ZZ#DCIM
Phone#ZZ#Downloads
Phone#ZZ#Pictures
```

__Desired Output__

```
Phone
  XX
    DCIM
    Downloads
    Pictures
  YY
    DCIM
    Downloads
    Pictures
  ZZ
    DCIM
    Downloads
    Pictures
```

Now since `folderList()` sorts the info, we might be able to introduce some logic to `syncthing/gui/default/index.html` that creates the sub-panels.

I have tried to template this out a few ways, but the tricky part is that each label is evaluated individually, making it difficult to determine the parent folder for a sub-group or where the group ends.

For example, if we evaluate `Phone#XX#DCIM` as a delimited label, we would want to:

- create a sub-panel for Phone
- create a sub-panel for XX
- create a button/div for DCIM

When the next entry comes up for `Phone#XX#Downloads`, ideally it would:

- skip sub-panel for Phone
- add to sub-panel for XX
- create a button/div for Downloads

Then on the 4th entry for `Phone#YY#DCIM`, it would:

- skip sub-panel for Phone 
- create a sub-panel for YY
- create a button/div for DCIM

But how can we know where the group starts or ends? Am I at least looking in the right direction?

**Comment by calmh** [comment]
First of all, I don't think arbitrary levels of grouping will work well in the current UI. I'd keep it to just headers and folders underneath, but that doesn't greatly affect the tech of it. I'd change `folderlist` to return a tree like structure:

```
[
  { header: "Phone#XX", children: [
    "Phone#XX#DCIM",
    "Phone#XX#Pictures",
  ] },
  { header: "Phone#YY", children: [
    "Phone#YY#DCIM",
    "Phone#YY#Pictures",
  ] }
]
```

and create two levels of loops in the UI HTML - an outer and an inner.

**Comment by whoabuddy** [comment]
@calmh That's a good point, I thought one or two levels made sense, but I can see how too many groups will start to look funny in the UI. Simple is always better.

I like the idea of returning a tree structure from `folderList`, but I am wrapping my head around some challenges for the GUI side of the loops since a folder group and a folder are fundamentally different.

Here is how I am understanding it:

__ng-repeat folder in folderList()__
- will return array of header/children
- the example below is a single folder and two groups

__Folder List__

```
Consulting
PhoneXX#DCIM
PhoneXX#Pictures
PhoneYY#DCIM
PhoneYY#Pictures
PhoneYY#Downloads
```

__folderList() result__

```
[
  { header: "Consulting", children: [
    "Consulting"
  ] }
  { header: "PhoneXX", children: [
    "PhoneXX#DCIM",
    "PhoneXX#Pictures",
  ] },
  { header: "PhoneYY", children: [
    "PhoneYY#DCIM",
    "PhoneYY#Pictures",
    "PhoneYY#Downloads",
  ] }
]
```

__if number of children = 1__
- create button for folder `data-target="#folder-X"`
  - add panel-progress
  - add panel-title
    - add panel-icon
    - add panel-status
    - add panel-title-text
- create div container for folder info `id="#folder-X"`
  - add Folder ID, Folder Path, etc.

__if number of children is > 1__
- create button for group `data-target="#group-X"`
  - add panel-title
    - add panel-title-text
- create div container for folders `id="#group-X"`
  - repeat for all children:
    - create button for folder `data-target="#folder-X"`
      - add panel-progress
      - add panel-title
        - add panel-icon
        - add panel-status
        - add panel-title-text
    - create div container for folder info `id="#folder-X"`
      - add Folder ID, Folder Path, etc.

Hopefully I am on the right track. This is definitely a bit over my head but the more I play with it the more it makes sense. I am new to Angular so this was a great way to dive in and learn more.

Also, I see that `buildTree` returns a similar type of output using a forward slash delimited value. Would that be an easier way to return the same type of formatted string without creating new code?

**Comment by hadn** [comment]
Thanks for working on this issue. It is also import to me. 

Conceptually, I believe that the file-folder managing system approach can be used to express the idea with a little change of names: 
 (1) A "group" is a entity in the system and consists of a name (given by a string), a reference to a "parent group" and possibly other attributes that could be defined later;
 (2) A main ("root") group exists by default and can not be deleted; 
 (3) When a group is created, it is linked either to the main group or to one of the existing groups specified by the user; 
 (4) Every folder should have a "group-in" attribute, for each it will be dynamically linked to during execution time; 
 (5) The attributes "group-parent" and "group-in" can be freely changed by the user, thus allowing to move folders and groups in the group hierarchy.  
 (6) When deleting a group, all groups and folders linked to it should automatically be linked to its parent group.

As described above, the syncthing group would be similar to a file-system folder, while a syncthing folder looks like a simple file. 

For the current time, this group structure would be mostly used for the visual organization of folders. However, later new features could be added such as allowing to share all folders in a group with a machine using a simple step in the graphical interface.

I also believe that it would be better to save the group structure (that is, its hierarchy) separated from the folder information and to link them dynamically when generating the GUI interface in run-time. But this is a software architecture decision that may be bounded by other constraints and requirements that I am not aware of.

Thanks again.

**Comment by jtagcat** [comment]
It should be made possible to also mass pause/resume/rescan the groups. I've splitted a huge directory to ~20 folders. (I have considered running multiple syncthing instances, but then it wouldn't benefit from the cross-folder reuse.)

**Comment by jasonnab** [comment]
Hi and early apologies for necrobumping. I am definitely interested in this becoming an included solution.

A conceptual idea I had would be to do the following:

- Run Syncthing as a docker/podman container
- Make use of a docker/podman-compose file
- A script can be written to start by doing one of two things: read from a specific file input in order to know who's folders belong to whom; alternatively, ask your clients to add specific ID or tag to their labels
- The script then stop the Syncthing container, move the folders to the appropriate organizational/client folder, and would read and add lines to the container compose file, mapping the now moved folders to where they should exist within the Syncthing container

This is obviously a bit ridiculous, but I wanted to share it as _in theory_ it could work...

Thanks

**Comment by CiAntRawknis** [comment]
For the time being, it be helpful to be able to sort based on device name

**Comment by Mikhael-Danilov** [comment]
This feature would be a lifesaver, for users with a lot of shared folders, like myself.

**Comment by KhariCode** [comment]
Judging by the age of this post there may not be action on it, but I too would also love the ability to Group Folders in some way, even if theres no way to mass pause and resume, it would be very useful in my case. I'm attempting to use Syncthing as my game backup solution between Steam Deck, PC & Synology. Right now I have like 25 folders and several of them could be nested into Steam Games , Emulation Saves & Projects etc

**Comment by acolomb** [comment]
Regarding storage of this information, I think it would be much easier to have a separate "group" attribute in the folder / device config, instead of trying to parse the label somehow.  The backend just hands out the value from the config file and doesn't handle it any other way.

The separator could be fixed to some character then.  But I question whether several levels of grouping really make sense?  If you allow arbitrary depth, the visual representation in the browser will get tricky, as it's not just a simple collapsible tree view.  Either way, each level should be collapsible, so it actually saves some screen real estate.  What status to show in the collapsed state needs to be discussed as well.

@dr460r would you be willing to contribute the actual code changes as well?

**Comment by dr460r** [comment]
> Regarding storage of this information, I think it would be much easier to have a separate "group" attribute in the folder / device config, instead of trying to parse the label somehow. The backend just hands out the value from the config file and doesn't handle it any other way.

If that's what maintainers think is a step in right direction for the project then sure. I'll tag @calmh since they were active maintainer in this conversation earlier. Also @lkwg82, they contributed label feature and suggested earlier using that for grouping.

The thing about using label is that backend doesn't need to be changed, and by using path-like naming convention people can opt-in to use grouping rendering and by doing that also have naming convention that is sort-friendly in GUI.

> If you allow arbitrary depth, the visual representation in the browser will get tricky, as it's not just a simple collapsible tree view.

Agree. @calmh was also talking about that earlier. One or two levels of grouping should work well in GUI, but if label is used to describe a path of groups, than that path could have any number of separators because only first 1 or 2 will be used to render 1 or 2 group levels, and rest of the path would then serve for sorting, but that's up to user if they want to use same naming convention for "grouping" part and for the rest of the label.

> would you be willing to contribute the actual code changes as well?

Sure, I would be interested in contributing in my spare time. However, I can only work on the web app side of things, since I don't know Go.

**Comment by acolomb** [comment]
> Sure, I would be interested in contributing in my spare time. However, I can only work on the web app side of things, since I don't know Go.

The needed changes to add a group field are really minor, because no additional logic is required.  Just need to adjust the protobuf definition and regenerate the implementation.  Don't hesitate to ask if you're having trouble with anything Go-related.

**Comment by opi3636** [comment]
just like @Dark1886 came here to say i would really appreciate this

**Comment by Hakker** [comment]
Same would love to see it as well mostly to keep it clean to look at. Now it can become quite the mess when you have a ton of folders.

The what and how to implement it I would look at far more intelligent people than me. I would only say keep it separate from the actual share so that even if the labels/tags whatever you just completely muck up then the share would just stay but then without a label/tag.

**Comment by acolomb** [comment]
> Sure, I would be interested in contributing in my spare time. However, I can only work on the web app side of things, since I don't know Go.

@dr460r Does this offer still stand?  I could help you out with the Go part, but I'm not that much into web design.  It should be pretty straight-forward, adding another one or two looping levels in our AngularJS template and wrapping inside the same elements used for the accordeon already.  You could start a PR with that (in draft state) and simply hard-code some group attribute into the folder config objects.  When the result looks acceptable, we can integrate that attribute into the backend.  Adding a configuration setting for the attribute would be the final step in the GUI again.

**Comment by superboo07** [comment]
Came here to say I'd also be heavily interested in this as I use syncthing heavily with my steam deck, laptop, and desktop. It would be huge to be able to categorize folders

**Comment by francescoragazzi** [comment]
+1 !

**Comment by foss-guru** [comment]
I was thinking of suggesting color coded folders or a tagging system, but this comes really close. That's a heck of a use case for sure to need 50 some groups. I think tagging or color coding would be good for advanced users that fall somewhere in the middle. I have 30-40 folder pairs on my main system at home clustered with various other devices each with about a dozen or so pairs. Tags or colors would be really helpful for keeping track of groups by environment. Windows, Linux, Android, etc...

**Comment by danielanywhere** [comment]
I am with @foss-guru. I would also be good with setting tags or colors on folders, then being able to filter by that. Filtering for tags/colors would be an easier to accomplish on the UI than full grouping layout, and the single-list filter results would be easier for a user to read.

**Comment by maen-bn** [comment]
👋  I've popped a PR in to provide a very simple start to this. I appreciate that it's not all singing and dancing but wanted to not over complicate it as this is my first contribution to syncthing and I'm not a UI person.

Here's how it's working

![Image](https://github.com/user-attachments/assets/b8d26dce-90fc-4989-8198-7ec119917d83)
